design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
+
+The following list of additional parameters will be supported:
+
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures.
```

Comment:
Is there a situation in which you would not want keycloak to fail its readiness check if the db was down? Should this just be a default behaviour (IE readiness check always fails if DB is inaccessible )

Reply:
Having in mind that we've decided to remove the External IdPs and User Federation from the Healthcheck API (see [this comment](https://github.com/keycloak/keycloak-community/pull/55#discussion_r498709614)), the database check is the only one that remains. I believe it makes a lot of sense to include it by default. Having in mind, that we perform a Passive checks, this is essentially free of charge for us.

So, yeah, totally agree.

Reply:
Added to the new version of the Design. Resolving.